OTEP: Span Event API deprecation plan (#4430)
open-telemetry/opentelemetry-specification: oteps/4430-span-event-deprecation-plan.md

# Span event deprecation plan

## Motivation

As stated in [OTEP 265: Event Vision](0265-event-vision.md),
the long-term plan is to deprecate span events in favor of
(log-based) events.

> 

Reviewer: I think we need to explain why we're deprecating them - I don't think we have ever done it. Suggesting to add some explanation like 

> Span events and log-based events both capture point-in-time telemetry. However, span events are limited because they need tracing instrumentation and can only be reported when a span ends. While both serve similar purposes, log-based events support a wider range of use cases.

Reviewer: I'm talking about the proto level - span events are exported along with spans.

Author: I like your proposed text, I've included it, thanks!

Since there are some backends that only support tracing and have no plans to
ingest log-based events directly, this statement has caused concern and worry
since there are no further details in that OTEP about how the deprecation
will work and whether these backends' needs will still be met by OpenTelemetry.

This OTEP lays out some requirements around the deprecation plan
to reduce confusion and hopefully alleviate these worries.

## Prerequisites

The prerequisites for the deprecation plan are:

> - The event specification has been stabilized.

Author: added link to https://github.com/open-telemetry/opentelemetry-specification/issues/4362

- There is [stable](../specification/versioning-and-stability.md#stable)
  support for (log-based) events in the relevant SDK.
- The [opt-in mechanism to emit (log-based) events
  as span events](#emitting-log-based-events-as-span-events)
  exists (and is stable) in the relevant SDK.

> ## Deprecation plan by component

Author: maybe a blog post where we describe the plan, and it can mention a tracking issue people can use to provide any feedback?

> ## Deprecation plan by component

Author: good idea, I added a section for communication plan

### Proto

Span events SHOULD be marked as deprecated in the proto definitions,
recommending that people use (log-based) events instead. Per
[OpenTelemetry proto stability rules](https://github.com/open-telemetry/opentelemetry-proto/blob/main/README.md#stability-definition)

> span events MUST never be removed from the proto.

Reviewer: nit, maybe this?

```suggestion
span events MUST NOT be removed from the proto.
```

(a slightly more nomative)

### Span event API

The [Span AddEvent](../specification/trace/api.md#add-events) API
SHOULD be marked as
[deprecated](../specification/versioning-and-stability.md#deprecated),
recommending that people use the OpenTelemetry Log (Event) API instead.
The default behavior of this method MUST NOT change.

The [Span AddEvent](../specification/trace/api.md#add-events) API
MAY only be [removed](../specification/versioning-and-stability.md#removed)
if the API (ever) bumps the major version
and the [opt-in mechanism to emit (log-based) events
as span events](#emitting-log-based-events-as-span-events)
does not rely on its presence.

The [Span RecordException](../specification/trace/api.md#record-exception) API
SHOULD be marked as
[deprecated](../specification/versioning-and-stability.md#deprecated),

> recommending that people use the OpenTelemetry Log API instead.

Reviewer 2: Please consider setting attributes on the span for unhandled exceptions as an alternative: https://github.com/open-telemetry/opentelemetry-specification/issues/4429#issuecomment-2681901775

Author: I've updated the OTEP to record span-terminating exceptions directly as span attributes

> The default behavior of this method MUST NOT change.

Reviewer 3: It's not clear to me that this can happen without a replacement. As discussed in https://github.com/open-telemetry/opentelemetry-specification/pull/4333#discussion_r1946331112 and some other spots I can dig up there is information transmitted by including exception details in a span that is not captured by log based events. Semantically it is useful to know if the span was ended by the exception or not (and if it was, which exception ended it) and in terms of information stack traces change as the well traverse the stack so the stack trace on each span would be different.

My feeling for where discussion on this topic is heading is that spans should have _attributes_ describing the _single_ exception that may have bubbled up through them including at least something like `exception.type` and `exception.messsage` and possibly also `exception.stacktrace` (configurable; users may be willing to forego that information for cost reasons).

There is more discussion in https://github.com/open-telemetry/opentelemetry-specification/issues/4429 and I do not mean to derail this PR but I did want to point it out and link the discussion.

Author: I've updated the OTEP to record span-terminating exceptions directly as span attributes

The [Span RecordException](../specification/trace/api.md#record-exception) API
MAY only be [removed](../specification/versioning-and-stability.md#removed)
if the API (ever) bumps the major version
and the [opt-in mechanism to emit (log-based) events
as span events](#emitting-log-based-events-as-span-events)
does not rely on its presence.

### Tracing SDK

The tracing SDK SHOULD provide an opt-in mechanism that allows

> users to emit span events as (log-based) events.

Reviewer: I wonder if we need this? We know there are backends that don't support logs, but do we have OTel backends that support logs, but not span events? 

I.e. would this be a concern at all if we let span events fade away organically?

This translation can also happen in the collector.

Reviewer 4: if a user/backend prefers to get Exceptions/events **consistently** on Logs, irrespective of instrumentation's adopting Log-based events, then users can leverage this.

Reviewer: I agree, the point I'm trying to make that it's an existing problem and it seems everyone who experienced it has solved it in some way. You can solve it in collector or on the ingestion side. I only doubt that it's a part of span event deprecation

Reviewer 3: It does seem like a reasonable separate request to me. It's also something that can be created as a 3rd party `SpanProcessor` and shared or something like that.

Author: I'll move this to a "future possibilities" section at the end.

Author: moved

This would be a short-term solution until existing instrumentations are
updated to emit (log-based) events.

### Instrumentation

For [stable](../specification/versioning-and-stability.md#stable)
instrumentations that are emitting span events:

- In the instrumentation's current major version
  - It SHOULD continue to emit span events.
- In the instrumentation's next major version
  - It SHOULD stop emitting span events and only emit log-based events.
    Users will be able to continue receiving span events by using the
    [opt-in mechanism to emit (log-based) events
    as span events](#emitting-log-based-events-as-span-events).

Non-stable instrumentations SHOULD use their best judgement on whether to follow
the above guidance.

> ## Emitting (log-based) events as span events

Author: I wonder if this is necessary if we go with https://github.com/open-telemetry/opentelemetry-specification/issues/4429#issuecomment-2682466455 (which would need its own backcompat story to emit terminal exceptions as span events instead of span attributes).

The Java instrumentation repo for example uses span events very rarely outside of terminal exceptions.

Reviewer 3: I do think that would be an simpler backward compat story because it doesn't rely on combining two signals or even different data points from the same signal! All of the data would be contained within the span itself.

I suspect very few users are calling `Span.RecordExcpetion` in Python manually -> I'd wager a bet that most span events have a single exception that is the terminal exception. I actually don't know any real world use cases that would result in a span would have multiple exception events but maybe I'm just not thinking of them.

Author: I've updated the OTEP to record span-terminating exceptions directly as span attributes and updated the backcompat story.

There are some backends that only support tracing and have no plans
to ingest log-based events.

To support this use case, there SHOULD be an opt-in mechanism that allows
users to emit log-based events as span events.

This opt-in mechanism SHOULD NOT be removed, even in a major version bump.

> This mechanism SHOULD be implemented as follows:

Reviewer 3: Yes I'd like to see this included in the SDKs and standardized as part of the migration process. We are lucky to be authors of a backend so I was able to add support for logs in a couple of hours but I know a lot of users of OpenTelemetry don't control their backends as much. For example [this comment](https://github.com/open-telemetry/opentelemetry-specification/issues/4414#issuecomment-2676297180):

> Please keep span events, and please allow us to control where event-like information is written. If that's logs by default, fine, but if I don't have a choice to keep that data in our tracing system that will make things extraordinarily difficult for us.

Personally I think we're going to run into a lot of users not wanting to use logs at all and it would have been a better route to transmit point in time information over the traces signal but that's a bigger topic.

> This mechanism SHOULD be implemented as follows:

Author: Included link to the Java prototype

- A log processor that converts events to span events and attaches them
  to the current span, whose behavior and configuration are defined

>   in the OpenTelemetry Specification.

Reviewer 5: Do we want to allow putting the bridged log records (without event name) as span events? I think it may be seen as benefit by users with no plans to ingest log-based events directly. Maybe as an opt-in feature as "Future possibility"?

Reviewer 3: It seems to me that would be going in the direction of keeping span events around forever.

I think one hard question that needs to be answered is what the story is for users with no plans to ingest logs in general (or at least not in the short term / it's hard for them). If we are saying span events go away we are basically leaving them with no options whereas the current status quo is that they are okay (right now you can use tracing without logs and mostly get all of the information, but after this change you'd have to be using deprecated span events to do so).

Author: Added under future possibilities

>   in the OpenTelemetry Specification.

Reviewer 3: Strongly agree. It's going to make the transition so much easier for users if they don't have to update their application code everywhere, especially in cases where they may not even control the application code (e.g. I deploy a Helm chart that thankfully can emit OTEL data... but they don't update their SDKs or don't offer the right hooks to configure this).

>   in the OpenTelemetry Specification.

Author: I've clarified that this is an SDK-based log processor, and added a Collector-based processor to the future possibilities. While I agree that would be really useful, I'm worried about the complexity and timeline if we introduce it as a blocker for Event stability, especially with the Collector folks head down on Collector 1.0 for the foreseeable future.

I'm resolving this discussion, but feel free to open a new discussion if you'd like to discuss further about whether it should be a blocker for Event stability.

- A standard way to add this log processor via declarative configuration

>   (assuming its package has been installed).

Reviewer 4: package installed + Logs SDK is setup/enabled.

Author: I think this is covered by "A standard way to add this log processor via declarative configuration", e.g. check out https://github.com/open-telemetry/opentelemetry-java-contrib/blob/80adbe1cf8de647afa32c68f921aef2bbd4dfd71/processors/README.md#event-to-spanevent-bridge

>   (assuming its package has been installed).

Reviewer 3: I agree it's complicated to do in the Collector but I think feasibility of doing so is important to decide what to do with exceptions and deprecation of span events. This same point was made [in another comment](https://github.com/open-telemetry/opentelemetry-specification/pull/4430/files#r1969066011).

The Collector is going to have to evolve to be able to manipulate Logs and Traces signals together and do the sort of buffering you are suggesting for multiple reasons:
- Backward compatibility configuration (this thread)
- Tail sampling (exceptions are only in logs)

>   (assuming its package has been installed).

Author: I've added a Collector-based processor to the future possibilities now. While I agree that would be really useful, I'm worried about the complexity and timeline if we introduce it as a blocker for Event stability, especially with the Collector folks head down on Collector 1.0 for the foreseeable future.

I'm resolving this discussion, but feel free to open a new discussion if you'd like to discuss further about whether a Collector-based processor should be a blocker for Event stability.

# Span event deprecation plan

## Motivation

[OTEP 265: Event Vision](0265-event-vision.md) states that we intend to
deprecate span events in favor of (log-based) events.

Span events and log-based events both capture point-in-time telemetry.
However, span events are limited because they need tracing instrumentation
and can only be reported when a recording span ends.
While both serve similar purposes, log-based events support a wider range of
use cases.

As we look at the current usage of span events though, by far the most common
use is for recording span-terminating exceptions, and it will be much less
disruptive to record these span-terminating exceptions directly as span
attributes (since there is only ever at most one of these) compared to
recording them using the Logs API:

- It is _much_ simpler to build a backcompat collector processor that moves
  span-terminating exceptions from span attributes to a span event,
  compared to buffering spans and moving these span-terminating exceptions

>   over to the appropriate span.

Reviewer: 🔥

- Some backends only support tracing and have no current plans to ingest
  log-based events directly.

## Prerequisites

> The prerequisites for deprecating span events are:

Author: I think this is the case for all oteps? also, I've restructured this part so ptal and see if it helps, thanks

> - The (log-based) event specification has been stabilized

Reviewer: Should we deprecate span events (in the spec) along with declaring events stable? 

in other words, this list talks about deprecation in specific API/SDK and it probably can only happen after they are deprecated on the spec level, so let's clarify when that happens?

  ([#4362](https://github.com/open-telemetry/opentelemetry-specification/issues/4362)).
- There is [stable](../specification/versioning-and-stability.md#stable)
  support for (log-based) events in the relevant SDK.
- The [backcompat story](#backcompat-story) exists (and is stable)

>   in the relevant SDK.

Author: I added this as a "Nice to have" at the bottom. I do agree we should do this, I'm just worried about getting any Collector resources due to the 1.0 effort.

  - It SHOULD continue to emit span events
    (continuing to call
    [Span RecordException](../specification/trace/api.md#record-exception)
    and [Span AddEvent](../specification/trace/api.md#add-events))
- In the instrumentation's next major version
  - It SHOULD stop emitting span events
    (instead calling Span "SetException" for span-terminating exceptions
    and calling the Logs API for all other use cases).
  - Users will be able to retain the old behavior by opting in to the
    [backcompat story](#backcompat-story).

Non-stable instrumentations SHOULD use their best judgement on whether to follow
the above guidance.

## Backcompat story

### Emitting span-terminating exceptions as span events via the SDK

This mechanism SHOULD be implemented as follows:

- An SDK-based span processor that converts span-terminating exceptions
  recorded as span attributes into span events.
- A standard way to add this span processor via declarative configuration
  (assuming its package has been installed).

Additionally, this span processor SHOULD be included in the standard
OpenTelemetry zero-code distribution (if one exists for the language)

> for at least 1 year.

Reviewer: nit:

I assume 1 year comes from the SDK/contrib stability guarantees, right? Isn't it redundant? I.e. processor is included into stable distro and can't be removed from it, and stable SDK/contrib has to be maintained for 1 year anyway. 

So a soft suggestion

```suggestion
```

Author: > I assume 1 year comes from the SDK/contrib stability guarantees, right?

are you're referring to https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#contrib-support?

that looks like it's about supporting a prior major version after a major version bump

Reviewer: yes, what I mean that we don't have per-feature support policy. If the processor is part of the stable distro, it has to stay there until the next major version (if ever). And if there is a new major version, the old one will be supported for 1 year.

So this requirement of 1 year is either impossible (if per-feature) or already covered by existing support policies.

Author: oh yes, I understand now, thanks, I will remove 👍

- A standard way to add this log processor via declarative configuration
  (assuming its package has been installed).

Additionally, this log processor SHOULD be included in the standard
OpenTelemetry zero-code distribution (if one exists for the language)
for at least 1 year.

### Emitting span-terminating exceptions as span events via the Collector

This mechanism SHOULD be implemented as follows:

- A Collector-based span processor that converts span-terminating exceptions
  recorded as span attributes into span events.

Additionally, this log processor SHOULD be included in the standard
OpenTelemetry Collector Contrib distribution for at least 1 year.

## Communication plan

Publish a blog post if/when this OTEP is accepted, giving readers a way to
provide feedback (e.g. pointing to a specification issue where we are
gathering feedback).

## Future possibilities

- An SDK-based log processor that converts all log records (not only event
  records) to span events and attaches them to the current span.
- An opt-in mechanism in the tracing SDK that allows users to emit span events
  as (log-based) events. This would only be a short-term solution until

>   existing instrumentations are updated to emit (log-based) events.

Author: let me know if you feel strongly about these future possibilities, I personally feel they are unnecessary and am thinking toward removing them so that people don't get the impression that the Log SIG is still planning to evaluate them / possibly work on them

Author: Removed these, but added a new "Nice to haves" section which are things that the Log SIG really should try to implement, but aren't strictly requirements.

4. Deprecate [Span AddEvent](../specification/trace/api.md#add-events),
   recommending instead that events are recorded using the Logs (Events) API.

   This can be done in parallel with 3.

### Per Instrumentation

For [stable](../specification/versioning-and-stability.md#stable)
instrumentations that are emitting span events:

- In the instrumentation's current major version
  - It SHOULD continue to emit span events
    (continuing to call
    [Span RecordException](../specification/trace/api.md#record-exception)
    and [Span AddEvent](../specification/trace/api.md#add-events))
- In the instrumentation's next major version
  - It SHOULD stop emitting span events
    (instead calling Span "SetException" for span-terminating exceptions
    and calling the Logs API for all other use cases).
  - Users will be able to retain the old behavior by opting in to the
    [backcompat story](#backcompat-story).

Non-stable instrumentations SHOULD use their best judgement on whether to follow
the above guidance.

## Backcompat story

### Emitting span-terminating exceptions as span events via the SDK

This mechanism SHOULD be implemented as follows:

> - An SDK-based span processor that converts span-terminating exceptions

Reviewer 6: Is this double publishing or it deletes from the source? For graceful transition double publishing is useful. We should probably support both options. Perhaps also make double publishing the default behavior for a period of time.

Author: added this:

> Users can choose to register a log exporter depending on if they also want to export event records as logs or not.

  recorded as span attributes into span events.
- A standard way to add this span processor via declarative configuration
  (assuming its package has been installed).

Additionally, this span processor SHOULD be included in the standard
OpenTelemetry zero-code distribution (if one exists for the language)
for at least 1 year.

### Emitting (log-based) events as span events via the SDK

This mechanism SHOULD be implemented as follows (see
[prototype](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/80adbe1cf8de647afa32c68f921aef2bbd4dfd71/processors/README.md#event-to-spanevent-bridge)):

> - An SDK-based log processor that converts event records to span events

Reviewer 6: Same question here: double publish?

Author: added this:

> Users can choose to register a log exporter depending on if they also want to export event records as logs or not.

  and attaches them to the current span, whose behavior and configuration
  are defined in the OpenTelemetry Specification.
- A standard way to add this log processor via declarative configuration
  (assuming its package has been installed).

Additionally, this log processor SHOULD be included in the standard
OpenTelemetry zero-code distribution (if one exists for the language)
for at least 1 year.

### Emitting span-terminating exceptions as span events via the Collector

This mechanism SHOULD be implemented as follows:

- A Collector-based span processor that converts span-terminating exceptions

>   recorded as span attributes into span events.

Reviewer 6: It may be worth also supporting the opposite transform (Span Events to Span Attributes), for compatibility scenarios with new backends and old SDKs.

# Span event deprecation plan

## Motivation

> [OTEP 265: Event Vision](0265-event-vision.md) states that we intend to

Reviewer 7: 265 says nothing of the sort. This PR is retroactively changing 265 to make it look like there was an approved decision to deprecate span events. Please keep this new OTEP independent and make arguments on their own merits.

Author: This is what 265 says:

> [Events are intended to replace Span Events in the long term. Span Events will be deprecated to signal that users should prefer Events](https://github.com/open-telemetry/opentelemetry-specification/blob/main/oteps/0265-event-vision.md#relationship-to-span-events)

Author: @yurishkuro I see that you updated your comment above, but I'm still not clear what you are asking for here

https://github.com/open-telemetry/oteps/pull/265 explicitly says "Span Events will be deprecated" and had 20 approvals

This OTEP is about the plan for how to get there.

Reviewer 7: That OTEP provides no justification for replacement either. I don't know how many people paid attention to the specific line you quoted - from my perspective any claim in an OTEP, especially one that calls for major break in backwards compatibility, is irrelevant if it does not provides cost/benefit analysis.

Reviewer: @yurishkuro is the ask to provide a clear justification or is there a push back on the direction?

Reviewer 7: I think the direction is feasible, but insufficiently documented, and as such unconvincing. I am all in favor of disagree & commit approach, but committing must be based on some arguments, not on a declaration "we will do X".

I would like to see:
  * motivation for a change
  * cost/benefit analysis
  * viable migration / backwards compatibility story

For instance, say a user used to log events to spans. After this change they could replace that logging with Log API and configure SDK to still send those events into spans. But that means all their other logging that previously didn't go to spans now would, potentially overloading their tracing backend.

There are likely many other edge cases like this. So I am not seeing how this change is so massively beneficial to a user. Instead of deprecating anything, the guidance could be "we recommend using Log APIs to record events", but still leave span events for situations where they make sense.

Reviewer 8: I have concerns that even if the spec carves out a space for these things to be emitted as span events that if that is actively discouraged that the SDKs will deprecate and remove that capability in order to reduce their surface area, and that backends will either deprecate and remove the capability or never invest into building that functionality in the first place.

It also feels like there should be an abstraction out in front that instrumentation uses to capture that things are happening in a consistent way, and configuration/convention used to bridge that into the telemetry signals that are emitted.  Otherwise these changes in guidance put a lot of onus onto instrumentation libraries to both follow the recommendations as well as provide suitable configuration.

Reviewer 9: > So I am not seeing how this change is so massively beneficial to a user. Instead of deprecating anything, the guidance could be "we recommend using Log APIs to record events", but still leave span events for situations where they make sense.

The question is, how valuable is it to the ecosystem to have a single stable and non-deprecated API to accomplish a particular thing? Having multiple stable ways of recording a piece of data pretty much guarantees fragmentation in instrumentation, and thus complicates the user story. With a single stable way to record a thing, there will of course be laggards / holdouts with the deprecated thing, but at least it envisions a future with consistency, and creates slightly stronger pressure to get there. 

> I have concerns that even if the spec carves out a space for these things to be emitted as span events that if that is actively discouraged that the SDKs will deprecate and remove that capability in order to reduce their surface area

Removal isn't allowed without a major version change to the API, which I do not want to entertain in any meaningful time frame.

Reviewer 10: > The question is, how valuable is it to the ecosystem to have a single stable and non-deprecated API to accomplish a particular thing? Having multiple stable ways of recording a piece of data pretty much guarantees fragmentation in instrumentation, and thus complicates the user story. 

To be clear, there is a huge user benefit to there being a single, well-documented, and universal 'Event' API rather than having two separate ones that work differently. Instrumenters shouldn't have to know or care if a user is using tracing (or if a trace was sampled out) and dual-write attributes to both an event API and a span event API. Having a single event API that the SDK can write out thru the log signal, embed into the span signal, or something else entirely strikes me as a clean separation of concerns and a benefit to both instrumenters, SDK authors, and end users.

Reviewer 3: I agree that having a single canonical way to do things is preferred; sometimes you've got to break some eggs to make an omelette.

The point that @HaloFour makes that I want to emphasize is:

> I have concerns that even if the spec carves out a space for these things to be emitted as span events that if that is actively discouraged that the SDKs will deprecate and remove that capability in order to reduce their surface area, and that backends will either deprecate and remove the capability or never invest into building that functionality in the first place.

My concern is that the discourse has been "this is not a breaking change, we won't remove them from the stable APIs so it will keep working" but it's likely that SDKs, implementers, etc. will slowly stop supporting it, or tools won't expose the right configurations to use the deprecated code paths. I would much rather the discourse were "this is a strong breaking change, let's evaluate the pros cons" with some mention of mitigation by deprecation vs. removal.

And like @yurishkuro I think a clear synthesis of what the pros of using logs over span events are. From recollection of discussion with other folks:
- Buffering of span events can be a problem (they can't be sent out until the span itself is sent out)
- Logs support complex attributes and spans don't (this is my opinion, I've been told by others they disagree this is an advantage)

Reviewer 10: > I share the concern is that the discourse has been "this is not a breaking change, we won't remove them from the stable APIs so it will keep working" but it's likely that SDKs, implementers, etc. will slowly stop supporting it, or tools won't expose the right configurations to use the deprecated code paths.

If the specification says that you must allow events to be recorded as span events, even if we deprecate the span event API, then why is this a concern? You're arguing that people will disobey the spec? If the rest of the project isn't faithfully implementing the specification we have bigger problems. :)

Author: hi folks!

I've updated the OTEP now to reflect this morning's productive spec meeting discussion about keeping span events at the proto layer to support use cases that rely on Span Events being emitted in the same proto envelope as their containing Span.

I'm going to resolve this long conversation, but please open new discussion(s) for any specific items that you believe are not yet sufficiently addressed.

Reviewer 3: FWIW I agree with all of the above that having a single canonical way to do things is a much better user experience, but it needs to be justified with pros/cons and frankly be mostly pros and almost no cons vs. the existing solution. I also agree that saying that the current solution will not be deprecated is a good way to keep the ball rolling.

Author: resolving this discussion with the following new language added to the OTEP (as discussed in today's spec meeting):

> After further discussions, we are only planning to deprecate the Span Event API, while retaining the ability to emit Span Events via the Events API.
>
> This achieves the primary goal, which is to provide a single consistent guidance to instrumentation authors to use the Event API, while still enabling use cases that rely on Span Events being emitted in the same proto envelope as their containing Span.

again, please go ahead and open new discussion(s) for any specific parts of this discussion that you believe are not yet sufficiently addressed

> deprecate span events in favor of (log-based) events.

Reviewer 5: Is there any place where it is described why we want to replace span events with event (log) records?

Reviewer 2: right below?

Span events and log-based events both capture point-in-time telemetry.
However, span events are limited because they need tracing instrumentation
and can only be reported when a recording span ends.
While both serve similar purposes, log-based events support a wider range of

> use cases.

Reviewer 5: Is there a place where the rationale for replacing span events with event (log) records is clearly explained in details? From what I see, the community is looking for more than just intent - specifically, the actual reasons behind this change.

So far, I’ve only come across [this PR](https://github.com/open-telemetry/opentelemetry-specification/pull/4333), which focuses on exceptions. What about regular events that aren’t exceptions?

> While both serve similar purposes, log-based events support a wider range of
use cases.

What use cases?

A clear explanation and analysis (e.g. pros and cons of both approaches) would be valuable for broader understanding. E.g. what about the fact that processing and exporting of span events can be faster as the data is aggregated? Or the lack of severity and complex types for span events?  Won't tail-sampling be more complex for event (log) records? Why those two mechanism cannot co-exist (why do we want to deprecate span events)? I think there are a lot of questions that should be an answered. The answers can be opinionated, but at least we should make our opinions clear and give the community a chance to provide feedback.

Author: This OTEP is about the deprecation plan. The deprecation decision was made in OTEP 265:

> [Events are intended to replace Span Events in the long term. Span Events will be deprecated to signal that users should prefer Events.](https://github.com/open-telemetry/opentelemetry-specification/blob/main/oteps/0265-event-vision.md#relationship-to-span-events)

That said, I do agree that a writeup of the rationale would be good as part of the communication plan. I've noted it below now in the communication plan section.

> As we look at the current usage of span events though, by far the most common

Reviewer 7: This claim appears to be unlikely because we lack a mechanism to "look at the current usage", since an unknown % of usage happens in proprietary code to which we don't have access. Has there been a documented large survey among end users? For example, we had a [case study](https://medium.com/jaegertracing/weave-trimmed-troubleshooting-fat-cut-api-response-time-from-seconds-to-milliseconds-with-jaeger-ab594c22c921) once where end users claimed they replaced most logging with writing events to Jaeger spans. 

Aside from proprietary usage, RPC streaming is a common place where span events are used even in oss instrumentation.

  in the relevant SDK.

## The plan

### Proto and specification

1. Introduce and stabilize Span SetException in the specification.

   This will allow recording span-terminating exceptions directly as span
   attributes, instead of recording them as span events.

   Stabilization will go through the normal process, requiring prototypes
   in at least three languages.

2. Stabilize (log-based) events in the proto and specification.

   This will allow recording events using the Logs API, instead of recording
   them as span events.

   This can be done in parallel with 1.

3. Deprecate
   [Span RecordException](../specification/trace/api.md#record-exception),
   recommending instead that span-terminating exceptions are recorded directly
   as span attributes via a new Span function "SetException", and recommending
   that other exceptions are recorded using the Logs API.

4. Deprecate [Span AddEvent](../specification/trace/api.md#add-events),
   recommending instead that events are recorded using the Logs (Events) API.

   This can be done in parallel with 3.

5. Deprecate span events in the proto, recommending the replacements mentioned

>    in 3 and 4 above.

Reviewer 5: What is the reason to deprecate it in the proto? Does it mean that the backends would not need to respect span events?

Author: in short, as OTEP 265 says, to signal that people should prefer (log-based) events

span events can never be removed from proto per proto stability rules

if we think deprecating them in proto gives the wrong impression to backends, we could remove this from the plan for now at least

Author: Resolving, I've update the otep to reflect this direction

  (assuming its package has been installed).

Users can choose to register a log exporter depending on if they also want
to export event records as logs or not.

This log processor SHOULD be included in the standard
OpenTelemetry zero-code distribution (if one exists for the language).

### Emitting span-terminating exceptions as span events via the Collector

This mechanism SHOULD be implemented as follows:

- A Collector-based span processor that converts span-terminating exceptions
  recorded as span attributes into span events.

This log processor SHOULD be included in the standard
OpenTelemetry Collector Contrib distribution.

## Communication plan

Publish a blog post if/when this OTEP is accepted, giving readers a way to
provide feedback (e.g. pointing to a specification issue where we are
gathering feedback).

## Nice to haves

- Make [emitting span-terminating exceptions as span events via the
  Collector](#emitting-span-terminating-exceptions-as-span-events-via-the-collector)
  a prerequisite for stabilizing Span SetException in any SDK.
- Forward compatibility story: A collector-based span processor that infers

>   if a span event represents a span-terminating exception, and adds it

Reviewer 2: BTW in theory the best way to determine this is by checking the `exception.escaped` attribute, except that this attribute got deprecated. And in Python at least it gets set to False incorrectly even when the span ends.

Author: yeah, I don't think other languages were even setting this at all

repositories though, by far the most common
use is for recording span-terminating exceptions, and it will be much less
disruptive to record these span-terminating exceptions directly as span
attributes (since there is only ever at most one of these) compared to
recording them using the Logs API:

- It is _much_ simpler to build a backcompat collector processor that moves
  span-terminating exceptions from span attributes to a span event,
  compared to buffering spans and moving these span-terminating exceptions
  over to the appropriate span.
- Some backends only support tracing and have no current plans to ingest
  log-based events directly.

## Prerequisites

The prerequisites for deprecating span events in the specification:

- The (log-based) event specification has been stabilized
  ([#4362](https://github.com/open-telemetry/opentelemetry-specification/issues/4362)).
- There is [stable](../specification/versioning-and-stability.md#stable)
  support for (log-based) events in the relevant SDK.
- The [backcompat story](#backcompat-story) exists (and is stable)
  in the relevant SDK.

## The plan

### Proto

> The span event proto will not be deprecated.

Reviewer 11: > in the proto, deprecated things can never be removed, but the concern is that new backends may not implement deprecated proto fields

+1
 
> in the SDK, deprecated things can be removed if there is (ever) a major version bump

Does this PR try to push for such removal + major version bump or not? I think folks normally worry less about "CAN", they care more about "SHOULD" and "MUST".

> The span event proto will not be deprecated.

Author: I can say confidently on the Java side there is zero interest in a major version bump ever

> The span event proto will not be deprecated.

Reviewer 11: Here is my suggestion:
1. If this PR is trying to push something to be removed at certain point in time, we should make it very clear in the doc.
2. If this PR is not trying to push something to be removed, then it is very much aligned with what @yurishkuro suggested here https://github.com/open-telemetry/opentelemetry-specification/pull/4430#discussion_r1976060830. In this case, I suggest that we either 1) avoid the "deprecation" word entirely (to avoid folks misinterpreting it) and stick with "we recommend that ...", or 2) we keep "deprecated/deprecation" wording and link to the clear definition/explanation.

> The span event proto will not be deprecated.

Reviewer 11: For example, if "deprecation" means "the doc is in a golden state and we expect no further updates" according to https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/document-status.md#lifecycle-status, I guess most folks won't have a problem.

> The span event proto will not be deprecated.

Author: I thought we solved this problem in the spec meeting today and it is reflected in the OTEP now?

> After further discussions, we are only planning to deprecate the Span Event
API, while retaining the ability to emit Span Events via the Events API.
>
> This achieves the primary goal, which is to provide a single consistent
guidance to instrumentation authors to use the Event API, while still
enabling use cases that rely on Span Events being emitted in the same proto
envelope as their containing Span.

> The span event proto will not be deprecated.

Reviewer 11: > I thought we solved this problem in the spec meeting today and it is reflected in the OTEP now?
> 
> > After further discussions, we are only planning to deprecate the Span Event API, while retaining the ability to emit Span Events via the Events API.
> > This achieves the primary goal, which is to provide a single consistent guidance to instrumentation authors to use the Event API, while still enabling use cases that rely on Span Events being emitted in the same proto envelope as their containing Span.

Yes, it is reflected in the OTEP now **in an unclear way** and that's why I asked the initial question. I think the problem is what we use the term "deprecate" without a good source of definition. Here is my suggestion:

Change the wording here, replace "The span event proto will not be deprecated" to "The span event proto will not be affected". And here is the reason:
1. We know what does "deprecate" mean for API/SDK https://github.com/open-telemetry/opentelemetry-specification/pull/4430#discussion_r1980350823
2. We know what does "deprecate" mean for telemetry signals, which is clarified here https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#deprecated
3. We know what does "deprecate" mean for proto, which is explained here https://github.com/open-telemetry/opentelemetry-proto/blob/main/README.md#experimental-additions-to-stable-components and it is only intended for Experimental fields - which won't apply here as we're talking about a stable field.

> The span event proto will not be deprecated.

Reviewer 11: ```suggestion
The span event proto will not be affected by this proposal.
```

> The span event proto will not be deprecated.

Author: thanks, makes sense, I started making this change, but then ended up improving the overall doc structure a bit and I think this section isn't needed anymore so removed it

### Specification

1. Introduce and stabilize Span SetException in the specification.

   This will allow recording span-terminating exceptions directly as span
   attributes, instead of recording them as span events.

   Stabilization will go through the normal process, requiring prototypes
   in at least three languages.

2. Stabilize (log-based) events in the proto and specification.

   This will allow recording events using the Logs API, instead of recording
   them as span events.

   This can be done in parallel with 1.

> 3. Deprecate
>    [Span RecordException](../specification/trace/api.md#record-exception),

Reviewer 11: The word "Deprecate" has many different meanings, would you update it to something like:

```suggestion
3. Mark
   [Span RecordException](../specification/trace/api.md#record-exception) as [Deprecated](../specification/document-status.md#lifecycle-status),
```

Author: good idea, added this in four places

The [Span AddEvent](../specification/trace/api.md#add-events) API
MAY be [removed](../specification/versioning-and-stability.md#removed)
if the API (ever) bumps the major version
AND if the [opt-in mechanism to emit (log-based) events
as span events](#emitting-log-based-events-as-span-events)
does not rely on it.

### Instrumentation

For [stable](../specification/versioning-and-stability.md#stable)
instrumentations that are emitting span events:

- In the instrumentation's current major version
  - It SHOULD continue to emit span events.
  - It SHOULD allow users to opt-in to emitting (log-based) events.
- In the instrumentation's next major version
  - It SHOULD stop emitting span events and only emit log-based events.
    Users will be able to continue receiving span events by using the
    [opt-in mechanism to emit (log-based) events
    as span events](#emitting-log-based-events-as-span-events).

Non-stable instrumentations SHOULD use their best judgement on whether to follow
the above guidance or not.

## Emitting (log-based) events as span events

There are some backends that only support tracing and have no plans
to ingest log-based events.

> To support this use case, there MUST be a simple opt-in mechanism that allows

Author: thanks! I incorporated most of this, let me know if I left out anything important / worth further clarification

for these config options:

> * send all logs as span events
> * send only log-based events as span events
> * when log is translated to span event, do we duplicate them or only send the span event

do you think we need them if we are only trying to solve this use case: "There are some backends that only support tracing and have no plans to ingest log-based events"?

> To support this use case, there MUST be a simple opt-in mechanism that allows

Reviewer: thanks, I think you covered everything! "send only log-based events as span events" seems to cover the disruptive part and the rest can be added in the future if necessary.
